### ISSUE TITLE
Test 85 failing under Windows in Debug configuration

### ISSUE THREAD

**Issue by claudiofantacci** [PointCloudLibrary/pcl#2570]
Here is the error:
```
85: Test command: C:\Users\cfantacci\GitHub\pcl\buildd\bin\test_registration.exe "C:/Users/cfantacci/GitHub/pcl/test/bun0.pcd" "C:/Users/cfantacci/GitHub/pcl/test/bun4.pcd" "C:/Users/cfantacci/GitHub/pcl/test/milk_color.pcd"
85: Test timeout computed to be: 9.99988e+06
85: [==========] Running 12 tests from 1 test case.
85: [----------] Global test environment set-up.
85: [----------] 12 tests from PCL
85: [ RUN      ] PCL.findFeatureCorrespondences
85: [       OK ] PCL.findFeatureCorrespondences (6 ms)
85: [ RUN      ] PCL.IterativeClosestPoint
85: [       OK ] PCL.IterativeClosestPoint (196 ms)
85: [ RUN      ] PCL.IterativeClosestPointWithRejectors
85: [       OK ] PCL.IterativeClosestPointWithRejectors (3641 ms)
85: [ RUN      ] PCL.JointIterativeClosestPoint
85: [       OK ] PCL.JointIterativeClosestPoint (24283 ms)
85: [ RUN      ] PCL.IterativeClosestPointNonLinear
85: [       OK ] PCL.IterativeClosestPointNonLinear (9110 ms)
85: [ RUN      ] PCL.IterativeClosestPoint_PointToPlane
85: [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
85: [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
85: [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
85: [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
85: [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
85: [       OK ] PCL.IterativeClosestPoint_PointToPlane (361 ms)
85: [ RUN      ] PCL.GeneralizedIterativeClosestPoint
85: [       OK ] PCL.GeneralizedIterativeClosestPoint (14878 ms)
85: [ RUN      ] PCL.GeneralizedIterativeClosestPoint6D
85: [       OK ] PCL.GeneralizedIterativeClosestPoint6D (9052 ms)
85: [ RUN      ] PCL.NormalDistributionsTransform
85: [       OK ] PCL.NormalDistributionsTransform (101659 ms)
85: [ RUN      ] PCL.SampleConsensusInitialAlignment
85: [       OK ] PCL.SampleConsensusInitialAlignment (47349 ms)
85: [ RUN      ] PCL.SampleConsensusPrerejective
85: [       OK ] PCL.SampleConsensusPrerejective (39103 ms)
85: [ RUN      ] PCL.PyramidFeatureHistogram
6/7 Test  #85: a_registration_test ..............***Failed  339.47 sec
```
![a_registration_test](https://user-images.githubusercontent.com/14168673/46945673-b0289200-d075-11e8-9dc5-3b019a338729.png)

This error is quite subtle and please let me know whether my understanding of the error, and its possible fix, is correct or not.

The main point of failure is this line here
https://github.com/PointCloudLibrary/pcl/blob/cd651056b0aa5f45bf23eeab6d81b2f42b18a009/registration/include/pcl/registration/impl/pyramid_feature_matching.hpp#L265
that pushes back into `access` variable an "unknown" number when `feature` is `nan`. Feature is never checked whether it could be `nan` or not and this may happen because of this snippet code from `PPFEstiamtion` feature used for the test (and maybe in user code also)
https://github.com/PointCloudLibrary/pcl/blob/cd651056b0aa5f45bf23eeab6d81b2f42b18a009/features/include/pcl/features/impl/ppf.hpp#L69-L118
variable `access` is then used to evaluate an index of a vector and since it results in `nan` its use is undefined (or platform dependent since it will be converted to a `size_t`).
I made some tests on some platform and apparently (take this with care, I'm not 100% sure) accessing a vector with `nan` results in accessing element at `0` (that is, the cast to `size_t` results in `0`).
It turns out that in `Debug` configuration under Windows, the index evaluated using `access` variable resulted in a very large number, not `nan`. As a consequence I got the out-of-bound access fualt from which I started the investigation.

Please note that if `access` is an arbitrary large number (as in `Debug` mode under Windows), the following line
https://github.com/PointCloudLibrary/pcl/blob/cd651056b0aa5f45bf23eeab6d81b2f42b18a009/registration/include/pcl/registration/impl/pyramid_feature_matching.hpp#L239
that translates `access` to a an index sometimes go over the maximum allowed representation of the type and, depending on the system, may result in a `0`, thus masking the unwanted behaviour.

The fix I propose, assuming that the expected behavior is to have `nan` features mapped to histogram bin `0`, is to change the following lines
https://github.com/PointCloudLibrary/pcl/blob/cd651056b0aa5f45bf23eeab6d81b2f42b18a009/registration/include/pcl/registration/impl/pyramid_feature_matching.hpp#L264-L265
in
```cpp
  for (size_t dim_i = 0; dim_i < nr_dimensions; ++dim_i)
  {
    if (!pcl_isnan(feature[dim_i]))
      access.push_back (static_cast<size_t>(std::floor((feature[dim_i] - dimension_range_target_[dim_i].first) / hist_levels[level].bin_step[dim_i])));
    else
      access.push_back (static_cast<size_t>(0));
  }
```
to force the inex to be `0` for `nan`s. By doing this the tests then run properly.

See further comments here https://github.com/PointCloudLibrary/pcl/issues/2399

**Comment by SergioRAgostinho** [comment]
I'm having trouble connecting all the dots. I need to have a look at the stack trace to be able to guide myself.

Regarding your proposal, it feels to me like we're hiding the problem under the carpet. I'm not fully into the problem like you, but intuitively it feels the problem needs to be fixed somewhere else. Inside the `at` method  I would simply throw I `feature[dim_i]` is `nan`.

**Comment by SergioRAgostinho** [comment]
Just want to quickly report my partial findings.

The first occurrence of nans in features happens on line 868
https://github.com/PointCloudLibrary/pcl/blob/50ca79cfc89a2e850ecdafb613b7a7bab6d8e7ca/test/registration/test_registration.cpp#L863-L868

The first and last points of the input cloud `ppf_signature_source` and ``ppf_signature_target` are set to NaN. The expressions below are being evaluated at line 298 
https://github.com/PointCloudLibrary/pcl/blob/50ca79cfc89a2e850ecdafb613b7a7bab6d8e7ca/registration/include/pcl/registration/impl/pyramid_feature_matching.hpp#L294-L299

```
(lldb) p input_->points.__begin_[0]
(pcl::PPFSignature) $13 = (f1 = NaN, f2 = NaN, f3 = NaN, f4 = NaN, alpha_m = NaN)
(lldb) p input_->points.__begin_[1]
(pcl::PPFSignature) $14 = (f1 = -0.0932385772, f2 = 0.00952175259, f3 = 0.0544939041, f4 = 0.00722551253, alpha_m = 2.07673001)
(lldb) p input_->points.__begin_[157607]
(pcl::PPFSignature) $15 = (f1 = 0.00231693685, f2 = 0.00789476931, f3 = -0.0350695662, f4 = 0.0119710751, alpha_m = 0.39232108)
(lldb) p input_->points.__begin_[157608]
(pcl::PPFSignature) $16 = (f1 = NaN, f2 = NaN, f3 = NaN, f4 = NaN, alpha_m = NaN)
```
I was able to verify the same with `ppf_signature_target`. 

My next steps would be to understand if:
1. these points can be filtered out safely. `ppf_signature_source` and `ppf_signature_target` don't have the same number of points so it might work.
2. understand the reason why it is apparently always the first and last point. I haven't confirmed if there are other points also set to NaN, but the first and last are consistently there.
3. does it make sense to let pyramid signature output NaNs. Is it trying to ensure it has the same number of points as the input cloud?

**Comment by claudiofantacci** [comment]
@SergioRAgostinho we are synchronized on the stack trace 👍 

I also noticed that the first and last element are always NaN, but I fear is something related to the test or on a common use case of the feature (evaluate the feature on a PointCloud without providing indices). If you look at how NaN are added
https://github.com/PointCloudLibrary/pcl/blob/cd651056b0aa5f45bf23eeab6d81b2f42b18a009/features/include/pcl/features/impl/ppf.hpp#L69-L118
it depends on the content of `indices` that cannot be known a-priori. As a result (by my understanding), there could be multiple NaN in the feature.

Let me reply in order:

> * these points can be filtered out safely. `ppf_signature_source` and `ppf_signature_target` don't have the same number of points so it might work.

I tried skipping NaN altogether in L298 of
https://github.com/PointCloudLibrary/pcl/blob/50ca79cfc89a2e850ecdafb613b7a7bab6d8e7ca/registration/include/pcl/registration/impl/pyramid_feature_matching.hpp#L294-L299
but the tests failed because the tolerance on the recognition tests were not respected.

My fear is that `dim_accumulator` in the following lines
https://github.com/PointCloudLibrary/pcl/blob/cd651056b0aa5f45bf23eeab6d81b2f42b18a009/registration/include/pcl/registration/impl/pyramid_feature_matching.hpp#L237-L243
is accomulated differently and something in the algorithm is altered for it to work properly.

That is the reason why I forced NaN to be pushed in bin 0, because it is actually what happens in the current implementation. 

> * understand the reason why it is apparently always the first and last point. I haven't confirmed if there are other points also set to NaN, but the first and last are consistently there.

My initial reasoning suggests that there may be other NaN points.

> * does it make sense to let pyramid signature output NaNs. Is it trying to ensure it has the same number of points as the input cloud?

I should read the paper of the subject feature to understand what is going on to propose a fix and some more detailed reasoning about this 😭 


Anyway, I'll try to skip again the NaN points to see whether I did something wrong during my first attempts and let you know. It is important however that we agree that there is an issue in the code.

**Comment by kunaltyagi** [comment]
> I made some tests on some platform and apparently (take this with care, I'm not 100% sure) accessing a vector with nan results in accessing element at 0 (that is, the cast to size_t results in 0).

Cast of floating NaN and Inf to integer as defined as UB. That means, we have to check the cast location and manually convert it to 0 in order to have consistent performance across platforms.

Though, @claudiofantacci mentioned trying to shoehorn NaNs into bin 0 already

**Comment by mvieth** [comment]
This problem was fixed by pull request #4711
